### PR TITLE
Fix cluster alert for watcher/monitoring IndexOutOfBoundsException

### LINKED ISSUE #43184: Watcher: getting painless script errors having null in email admin watches
**Elasticsearch version** (`bin/elasticsearch --version`): 7.1.1
**Plugins installed**: no
**JVM version** (`java -version`): On Elastic Cloud

**OS version** (`uname -a` if on a Unix-like system): On Elastic Cloud

**Description of the problem including expected versus actual behavior**: 
I thought this has been fixed in https://github.com/elastic/elasticsearch/pull/32923.
But I noticed a cluster running 7.1.1 is still getting the following error:
```
[instance-0000000006] failed to execute [script] transform for [h0rBJUPOSgeB-ZGRbrQ60A_elasticsearch_cluster_status_77a4d766-3da2-4b2b-92d9-451de304e9cf-2019-06-13T02:31:50.556Z]
org.elasticsearch.script.ScriptException: runtime error
	at org.elasticsearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:94) ~[?:?]
	at org.elasticsearch.painless.PainlessScript$Script.execute(ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ct ...:1152) ~[?:?]
	at org.elasticsearch.xpack.watcher.transform.script.ExecutableScriptTransform.doExecute(ExecutableScriptTransform.java:49) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.xpack.watcher.transform.script.ExecutableScriptTransform.execute(ExecutableScriptTransform.java:38) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.xpack.watcher.transform.script.ExecutableScriptTransform.execute(ExecutableScriptTransform.java:23) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.xpack.watcher.execution.ExecutionService.executeInner(ExecutionService.java:505) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.xpack.watcher.execution.ExecutionService.execute(ExecutionService.java:309) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.xpack.watcher.execution.ExecutionService.lambda$executeAsync$5(ExecutionService.java:410) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.xpack.watcher.execution.ExecutionService$WatchExecutionTask.run(ExecutionService.java:605) [x-pack-watcher-7.1.1.jar:7.1.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) [elasticsearch-7.1.1.jar:7.1.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653) ~[?:1.8.0_144]
	at java.util.ArrayList.get(ArrayList.java:429) ~[?:1.8.0_144]
	at org.elasticsearch.painless.PainlessScript$Script.execute(ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ct ...:429) ~[?:?]
	... 11 more
```

**Steps to reproduce**:

The admin email has not been set up. 
And just having the default system watchers.

### PR DESCRIPTION
**Pull request by jakelandis**
If there is an un-resolved cluster alert, but there is no .monitoring-es
data for this cluster in the last 2 minutes a run time error in the
Watcher transform can occur [1]. This could be due some slow down in
getting the .monitoring-es data from source cluster to monitoring cluster.

Prior to this commit, the Watch condition can pass if there are un-resolved
alerts, but the cluster state could not be found. Since the transform assumes
there is a cluster state (green,yellow,red) and could throw an error while
trying to read it.

This commit prevents passing the condition if the cluster state was not
found. 

Fixes #43184

[1]
```
org.elasticsearch.script.ScriptException: runtime error
	at org.elasticsearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:94) ~[?:?]
	at org.elasticsearch.painless.PainlessScript$Script.execute(ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ct ...:1152) ~[?:?]
	at org.elasticsearch.xpack.watcher.transform.script.ExecutableScriptTransform.doExecute(ExecutableScriptTransform.java:49) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	...
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
```
from .watcher-history:
```
     "type" : "script_exception",
                "reason" : "runtime error",
                "script_stack" : [
                  "java.util.ArrayList.rangeCheck(ArrayList.java:657)",
                  "java.util.ArrayList.get(ArrayList.java:433)",
                  "state = ctx.payload.check.hits.hits[0]._source.cluster_state.status;",
                  "                                   ^---- HERE"
                ],
```
---------
Screen shot of the formatted diff:
image

EDIT: changed the implementation to prevent passing the condition.

### CODE REVIEW

**Review by pickypg — APPROVED** [review]
LGTM

File: x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json [hunk 1]
```diff
@@ -134,7 +134,7 @@
...
...email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ctx.vars.is_new = ctx.vars.fails_check && !ctx.vars.not_resolved;ctx.vars.is_resolved = !ctx.vars.fails_check && ctx.vars.not_resolved;def state = ctx.payload.check.hits.total == 0 ? 'unknown' : ctx.payload.check.hits.hits[0]._source.cluster_state.status;if (ctx.vars.not_resolved){ctx.payload = ctx.payload.alert.hits.hits[0]._source;if (ctx.vars.fails_check == false) {ctx.payload.resolved_timestamp = ctx.execution_time;}} else {ctx.payload = ['timestamp': ctx.execution_time, 'metadata': ctx.metadata.xpack];}if (ctx.vars.fails_check) {ctx.payload.prefix = 'Elasticsearch cluster status is ' + state + '.';if (state == 'red') {ctx.payload.message = 'Allocate missing primary shards and replica shards.';ctx.payload.metadata.severity = 2100;} else {ctx.payload.message = 'Allocate missing replica shards.';ctx.payload.metadata.severity = 1100;}}ctx.vars.state = state.toUpperCase();ctx.payload.update_timestamp = ctx.execution_time;return ctx.payload;"
```
Comment (ycombinator) [inline x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json]: The addition of the `ctx.payload.check.hits.total == 0` check looks good to me but I'm wondering if `unknown` is the best state to use here. Why not leave the alert in it's current state?

**Comment by jakelandis** [comment]
@ycombinator @pickypg - thanks for taking a look.  In additional testing, I found that introducing the "unknown" state incidentally allowed for watch to incorrectly resolve itself if no data was found in past 2m. The exception prevented this from happening, but by fixing the exception, it introduced the possibility for an incorrect resolution. 

I changed the implementation (entirely) and now no longer allow the condition to pass if the cluster state can not be found.

File: x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json (hunk 1 above)
Comment (jakelandis) [reply x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json]: I changed the implementation to no longer change the state: https://github.com/elastic/elasticsearch/pull/45294#issuecomment-519199700

**Comment by jakelandis** [comment]
...ug... hold off on reviewing , I noticed the UI is now incorrectly claiming "OK" now, even though the email action is not firing :(

**Comment by jakelandis** [comment]
closing this PR.. will open a new one here in a bit with a more comprehensive fix. 

* Can not change the condition since it impacts the "OK" vs. "firing" in the UI
* Can not simply introduce an unknown state since it can fire an email after 2m of no-data with an incorrect "resolution" 
* Can not add unknown state check to the email action since the index action still fires, resulting in a missing email action. 

I think I need to a) not change the condition b) account for the missing hit array when trying to read it c) don't fire any actions if the state can not be found. 

To avoid more confusion, I will open a new PR with these changes.

**Comment by jakelandis** [comment]
Sorry for the churn ... I _think_ I have this figured out over on https://github.com/elastic/elasticsearch/pull/45308

**Pull request closed without merging**